src/chains/ethereum/ethereum/src/api.ts

```diff
@@ -2782,6 +2782,169 @@ export default class EthereumApi implements Api {
+   * @example for: blockCount: 2, newestBlock: 0x10, rewardPercentiles: [10, 25, 50]
+   * {
+   *   oldestBlock: "0x10",
+   *   reward:[["0x59682f00","0x2a7b749c11","0x2a7b749c11"]],
+   *   baseFeePerGas:["0x40c9459e7","0x420068cc1","0x420068cc1"]
+   *   gasUsedRatio: [0.5750394293729167, 0.50]
+   * };
+   *
+   */
+  @assertArgLength(3)
+  async eth_feeHistory(
+    blockCount: QUANTITY,
+    newestBlock: QUANTITY | Ethereum.Tag,
+    rewardPercentiles: number[]
+  ): Promise<Ethereum.FeeHistoryResult> {
+    let reward = [];
+    const baseFeePerGas = [];
+    const gasUsedRatio = [];
+    const blockchain = this.#blockchain;
+    const PAD_PRECISION = 16; // Matching geth/infura decimal precision formatting
+    const PRECISION = 10 ** PAD_PRECISION;
+    const PRECISION_BIG_INT = BigInt(PRECISION);
+    const PRECISION_BIG_INT_PERCENTILE = PRECISION_BIG_INT * 100n;
+
+    let blocksToFetch = Quantity.toNumber(blockCount);
+    let newestBlockNumber = blockchain.blocks
+      .getEffectiveNumber(newestBlock)
+      .toNumber();
+    let oldestBlock;
+
+    let currentBlockNumber = newestBlockNumber;
+
+    while (blocksToFetch > 0 && currentBlockNumber >= 0) {
+      const currentBlock = await blockchain.blocks.get(
+        `0x${currentBlockNumber.toString(16)}`
+      );
+      oldestBlock = currentBlockNumber;
+
+      const gasUsed = currentBlock.header.gasUsed.toBigInt();
+      const gasLimit = currentBlock.header.gasLimit.toBigInt();
+      const baseFee = currentBlock.header.baseFeePerGas
+        ? currentBlock.header.baseFeePerGas
+        : "0x0"; // needed to match infura. Genesis blocks without EIP-1559 (e.g. forking mainnet) will be null without this.
+
+      // baseFeePerGas
+      if (currentBlockNumber === newestBlockNumber) {
+        baseFeePerGas.unshift(
+          Quantity.from(Block.calcNextBaseFee(currentBlock))
+        );
+      }
+      baseFeePerGas.unshift(baseFee);
+
+      // gasUsedRatio
+      if (gasUsed === 0n) {
+        gasUsedRatio.unshift(0);
+      } else {
+        gasUsedRatio.unshift(
+          Number(
+            `0.${((gasUsed * PRECISION_BIG_INT) / gasLimit)
+              .toString()
+              .padStart(PAD_PRECISION, "0")}`
+          )
+        );
+      }
+
+      // reward percentile
+      if (rewardPercentiles.length > 0) {
+        const transactions = currentBlock.getTransactions();
+
+        // If there are no transactions, all reward percentiles are 0.
+        if (transactions.length === 0) {
+          reward.unshift(
+            rewardPercentiles.map(() => {
+              return "0x0";
+            })
+          );
+        } else {
+          // Get receipts
+          const transactionReceipts = await Promise.all(
+            transactions.map(async (t, i) => {
+              return (
+                await blockchain.transactionReceipts.get(t.hash.toString())
```

Comment:
```suggestion
                await blockchain.transactionReceipts.get(t.hash.toBuffer())
```
Prefer `.toBuffer()` here because `.toString()` has to convert to string and back to buffer.

Reply:
:+1: https://github.com/trufflesuite/ganache/pull/3351/commits/126b8e0373afceec46361ddbbd45973819223381